net/net_evbuffer.c

```diff
@@ -1551,11 +1561,45 @@ static int process_payload(struct event_info *e)
     }
 }
 
+static void evbuffer_free_callback(const void *ptr, size_t unused0, void *unused1)
+{
+    free((void *)ptr);
+}
+
 static ssize_t readv_evbuffer(struct evbuffer *buf, int fd)
 {
 #   define NVEC 8
     struct iovec v[NVEC];
-    const int nv = evbuffer_reserve_space(buf, MB(4), v, NVEC);
+    int avail = MB(4);
+
+    /*
+       The goal is to read as much as possible, up to 4MB. Previously we used
+       evbuffer_reserve_space(size=4MB,nvecs=8)/evbuffer_commit_space().
+       evbuffer_reserve_space() allocates memory when needed. The nvecs arugment
+       is just a hint: a large allocation will likely end up in a single iovec.
+       evbuffer_commit_space() pins the iovec structures that data is read into.
+
+       On a congested network this approach ends up wasting a lot memory.
+       A network message may take quite a few events to read completely
+       (a blob operation, for instance). The 1st event asks for 4MB, gets a
+       single 4MB iovec, and reads only a couple hundred of bytes into the 4MB
+       iovec. The 2nd event asks for another 4MB. It can't use the memory from
+       the previous iovec because it's pinned. So it gets another 4MB iovec,
+       reads another couple hundred bytes, and pins the entire iovec again.
+       And so on so forth...
+
+       Instead, we first determine how many bytes are immediately available for reading in
+       the fd, via ioctl(FIONREAD). We then reserve precisely this many bytes in the evbuffer,
+       readv, and commit the space.
+     */
+
+#ifdef FIONREAD
+    if (ioctl(fd, FIONREAD, &avail) == 0) {
+        if (avail > MB(4)) /* limit the rate */
```

Comment:
Let us read as much as possible without capping. In fact, 7.0 defines TCPBUF at 8MB. So at least that much?